### ISSUE TITLE
webgl is disabled in chromium webvr openvr build

### ISSUE THREAD

**Issue by vt5491** [toji/chrome-webvr-issues#52]
I am trying to get an openvr (steamvr) enabled OSVR device to use webvr as suggested in [this link](https://www.reddit.com/r/OSVR/comments/46r8h8/i_got_webvr_working_with_my_hdk_uses_steamvr/).  I am using the  'chromium_webvr_v1_win_openvr.7z' build of 3/7  from Brandon Jones [on google drive. ](https://drive.google.com/folderview?id=0BzudLt22BqGRbW9WTHMtOWMzNjQ#list)

My code is failing when attempting to get a webgl context:

```
this.canvas.getContext('2d')
->CanvasRenderingContext2D {canvas: canvas#viewer.keypress-events, globalAlpha: 1, globalCompositeOperation: "source-over", webkitImageSmoothingEnabled: true, imageSmoothingEnabled: true…}
this.canvas.getContext('webgl')
->null
this.canvas.getContext('experimental-webgl')
->null
this.canvas.getContext('experimental-webgl2')
->null
this.canvas.getContext('bitmaprenderer')
->null
```

As you can see, only the 2d context is set.  

The [webgl test page](http://get.webgl.org/) says:

`Hmm. While your browser seems to support WebGL, it is disabled or unavailable. If possible, please ensure that you are running the latest drivers for your video card.`

I am using an nvidia GeForce 650m graphics card with driver version 362.00.

I have tried enabling all the following under chrome://flags to no avail:

```
> WebGL Draft Extensions (#enable-unsafe-es3-apis)
> WebVR (#enable-webvr)
> Override software rendering lisl (#ignore-gpu-blacklist)
```

What do I need to do to get webgl enabled on this VR build?

**Comment by vt5491** [comment]
Interestingly, when I clicked the "Reset to Default" button, and relaunched, then webgl started working.  However, webvr is not enabled by default.  So then I enabled webvr (#enable-webvr), and relaunched.  Then I was back in the original problem of 'no webgl'.

So activating 'web-vr' seems to deactivate the webgl support somehow.

**Comment by toji** [comment]
First off, you can only get a single type of context from a canvas. So if you call `canvas.getContext('2d')` and then later on that same context call `canvas.getContext('webgl')` only the first one will succeed. Likewise if you get a `webgl` context first you won't be able to get a `2d` context later. You'll have to create a new canvas for each.

That doesn't sound like the root of your problem, though. I would recommend that while trying to get something like this running you *don't* use `--enable-unsafe-es3-apis` because that's still a work in progress and has had known stability issues. After ensuring that's off I would visit a page like the WebGL test page you linked earlier and, if it's still failing, navigate to `about:gpu` and attach the contents of that page here so I can help diagnose the issue further.

**Comment by vt5491** [comment]
Ok, that's good to know about getting multiple contexts -- I was not aware of that, although the lines I provided were just me typing in commands at the javascript console not the actual code that's failing.  You're right that my example doesn't really illustrate the problem, as the successful 2d call would preclude the subsequent commands from working.  

The actual line of code that's failing is in three.js  at THREE.WebGLRenderer: 

```
		_gl = _context || _canvas.getContext( 'webgl', attributes ) || _canvas.getContext( 'experimental-webgl', attributes );

		if ( _gl === null ) {

			if ( _canvas.getContext( 'webgl' ) !== null ) {

				throw 'Error creating WebGL context with your selected attributes.';

			} else {

				throw 'Error creating WebGL context.';

			}

		}

```
The problem is still occurring with just the 'enable-webvr' set.  I'm attaching files and screen prints of 'about:gpu' from a chromium 51 with webgl not working (enable-webvr set), and one from a chromium 51 with the webgl working (enable-webvr disabled).

Can I take it take it that you are unable to reproduce the problem on your end?  It should be easy to test, just download the build, enable/disable webvr and go to http://get.webgl.org/

I did do a vimdiff on the two files.  Here is a screen print of an interesting section.  The nowebgl is on the left, and the working webgl is on the right.  I notice how on the working version (including my standard chrome), the "Pixel Shader" version is 5.0, and GL_VENDOR is "google", whereas on the no webgl it's "Pixel Shader" 4.0, and GL_VENDOR is "Intel".  Don't know if this would be causing the problem, but it's about the only thing that jumped out at me.

![image](https://cloud.githubusercontent.com/assets/8662470/13721209/9d33580c-e7d3-11e5-8b05-9204d338aac8.png)

Here is the no_webgl version:
[chromium_51_nowebgl.txt](https://github.com/toji/chrome-webvr-issues/files/170280/chromium_51_nowebgl.txt)

and here is the working version:
[chromium_51_webgl_working.txt](https://github.com/toji/chrome-webvr-issues/files/170281/chromium_51_webgl_working.txt)

I am running windows 10, HP laptop.  Here is some more gpu info:
![image](https://cloud.githubusercontent.com/assets/8662470/13721240/7bc7d55c-e7d4-11e5-9072-9ca60743ba94.png)

Thank you for your assistance on this matter.

**Comment by vt5491** [comment]
I upgraded my nvidia to 364.51.  Problem still occurs.

**Comment by vt5491** [comment]
I was able to get this to work by right-clicking on the chromium binary, selecting "run with graphics processor->High performance nvidia processor".

Now my about: gpu shows:
GL_VENDOR	NVIDIA Corporation
GL_RENDERER	GeForce GT 650M/PCIe/SSE2
GL_VERSION	4.5.0 NVIDIA 364.51

I referred to [this chrome ticket](https://bugs.chromium.org/p/chromium/issues/detail?id=476203) which tipped me off to this.  In short, I think this is a chromium thing and has nothing to do with webvr.  The intel 4000 GL_RENDERER that was being used in the failure case, is the internal graphics card on the motherboard.  I guess it's just not powerful enough to run webgl.  Anyway, it's a hardware detection problem of chromium that appears to be activated when you select "enable vr".

I was trying some of the other builds (which is kind of expensive because I find that if you bring up version 51.0 and then try to start, say, version 48.0 the help still shows "51.0"..In other words I think DLLs are getting registered for the first version, and you really need to cycle your system to make sure you're getting a clean test).  Anyway, the last build I could get webgl working out of the box was the oculus build from 4/26/15.  I either was not getting webgl to work, or failing to start with message "[7936:9744:0314/221622:FATAL:extension_web_ui.cc(58)] Check failed: false.".  

In short, the webvr builds seem to be hit or miss with me, and it doesn't seem to be tied to just this one version, which is what made me start to investigate it as chromium problem and not a webgl or webvr problem.

Anyway, I'm happy now that I have a workaround.  Currently, I have updated my webapp to use webvr 1.0 protocols (using the [webvr-boilerplate v 1.0 tree](https://github.com/borismus/webvr-boilerplate/tree/webvr-1.0)), and I am getting head tracking to work on OSVR HDK 1.3.  I am not getting split screen just yet, but that's presumably not related to this issue.

I'll keep this open for the time while I continue to test, but I think this problem has more or less been resolved.  

Thank you for tip on the 'about:gpu', which helped my narrow in on the problem.
